core/commands/pin.go

```diff
@@ -377,13 +378,18 @@ new pin and removing the old one.
 			return
 		}
 
-		fromc, err := core.ResolveToCid(req.Context(), n, from)
+		r := &path.Resolver{
```

Comment:
At the end of the day, I think a better solution would be to avoid this resolver dance altogether. The root of the issue comes from the fact that the path doesn't dictate how it should be interpreted: as an IPLD path through IPLD objects (where the segments are keys) or an IPFS path through potentially sharded IPFS directories (where the segments are directory entries). To work around this:

1. It's up to commands to chose the initial resolver.
2. Resolvers do some additional magic under the covers to detect if a node should be treated as an IPLD or IPFS object by looking at it's Cid (they *really* shouldn't be doing this).

The solution I've proposed is to use separate prefixes: `/ipld/...` and `/ipfs/...` (https://github.com/ipfs/notes/issues/248). I believe this is how this was *intended* to work (when designed) anyways.

If we did this,  we could pass around the protocol along with the Cid/relative path so that `core.Resolve` and `core.ResolveToCid` could pick the right resolver for the given protocol.